### DISCUSSION TITLE
gravity magnitude jumps around sidesets and lower dimensional block

### DISCUSSION THREAD

**Discussion by Rahim-Habibi** [idaholab/moose#26670] [Q&A Modules: Porous Flow]
Hi all, 
 I am working on thermo-hydro-mechanical geothermal reservoir simulation, I want to model the response of a fault against water injection. To model the 2D fault i have to option 1) defining fault as a physical entity using Gmsh along with "lower_dimensional_block" and 2) defining fault as a lower dimensional block using MOOSE options (i.e.  SideSetsBetweenSubdomainsGenerator and LowerDBlockFromSidesetGenerator). So, the input file for both is the same only mesh block changes. 
the model has 15-15-6 km dimension in x,y and z direction and the fault located at the center. Accoring to the manual of "EquelValueConstraint" [LINK ](https://mooseframework.inl.gov/source/constraints/EqualValueConstraint.html), I defined two sidessets on the boundaries of the subdomains at two sides of the fault as follows:
```
[Mesh]
  [./file]
    type = FileMeshGenerator
    file = tpv104-inbox-fault.msh
  [../]
  [./sidesetmid1]
    type = SideSetsBetweenSubdomainsGenerator
    input = 'file'
    primary_block = 'SDMidBack'
    paired_block = 'SDMidFront'
    new_boundary = 'primary1_interface'
  [../]
  [./sidesetmid2]
    type = SideSetsBetweenSubdomainsGenerator
    input = 'sidesetmid1'
    primary_block = 'SDMidFront'
    paired_block = 'SDMidBack'
    new_boundary = 'primary2_interface'
  [../]
  [interfacemid]
    type = LowerDBlockFromSidesetGenerator
    input = sidesetmid2
    sidesets = 'primary2_interface'
    new_block_id = '10000'
    new_block_name = 'interfacemid'
    []
    []
```

Here is the model (figure 1):
![image](https://github.com/idaholab/moose/assets/57484835/8c393f89-6b89-4014-a597-d54653afb184)

When I am using the first technique (adding "lower_dimensional_block" phrase in .msh mesh file), everything is fine and the outputs make sense. 

However, I have to use second technique as I need to have "EqualValueConstraint" in my simulation meaning that I have to have sidesets and LDBlock in the mesh. So, I defined then as above. At this step, I only defined gravity in the model. But the stress distribution along depth seems wrong. in the following, more detials are given: 
According to Mechanics, I tried to set gravity and I applied it using two methods available in Porous Flow including : Gravity Kernel and Action. I drew a line crossing the center of the model from top to bottom to check gravity distribution over depth. But at the bottom of the fault I have got strange jump in gravity (figure below, figure 2) at two scenarios, kernel and action:
![image](https://github.com/idaholab/moose/assets/57484835/a0d0f520-e548-4b95-98a3-1237c9556352)

Also, as you can see, displacement seems strange as well, i.e. lower dimensional block OR sidesets behave like sink where the magnitude differs alot w.r.t. around. 
Just to make a comparisson, i plotted the distribution at a point far away from the lower dimensional block and sidesets. It is a bit shocking, since the jump disappears. Also, comparing this figure with the figure above (figure 2), it can be see that the gravity magnitude changes at the top alot. in this figure, the magnitude starts from about zero which makes sense BUT in the figure above it starts from positive and drops suddenly to negative magnitude. 

![image](https://github.com/idaholab/moose/assets/57484835/fe61bbe4-9e4f-4292-8eed-d5509b8dbec1)
 To delve more into the issue, i clipped the model at the center where fault exsits to see the distribution. It is obvious that the magnitude along the fault changes alot (and does not make sense).
![image](https://github.com/idaholab/moose/assets/57484835/f678c7fc-e7db-4bdb-b1d5-b786dc28ff2f)
I can not understand what is happening once lower dimensional block OR sidesets are working with 3D elements. How can i solve this issue? 

BTW, I got the almsot same result for the gravity distribution when I set hydrostatic stress state in the model, Sxx=Syy=Szz. 

Thank you in advance

**Comment by GiudGiud** [comment]
Hello

I wonder if the gravity kernel is not applied on all the elements it should. That could cause artificial stress.
@1runer thoughts?

Guillaume

**Comment by Rahim-Habibi** [comment]
Hi, 
@GiudGiud  and @cpgr 
I thought, the issue comes from my mesh or input file etc, so I tried one of the tests from MOOSE which has lower dimensional block([(modules/tensor_mechanics/test/tests/j_integral/j_integral_2d_block_restrict.i)](https://mooseframework.inl.gov/source/meshgenerators/LowerDBlockFromSidesetGenerator.html#4da9d292-5982-4d69-a49a-5f56a6163ccd)), without any edit in input file I just plot the stress along a line, the following show stress-xx and stress-yy along a line:
![image](https://github.com/idaholab/moose/assets/57484835/a60828a4-0a1f-4d7d-aca2-14d2d9d00964)
![image](https://github.com/idaholab/moose/assets/57484835/62fc2fba-dcfb-46f2-b871-47446170e29d)

So, as you can see the jump shows up on lower dimensional block in this example either stress-xx or stress-yy as well.
why such jump happens?
Thanks in advance
Rahim

**Reply by cpgr** [reply]
I wonder if it is just an output artefact that the stress auxvariables aren't projected to the lower dimensional block?

**Reply by GiudGiud** [reply]
I would expect a 0 in that case

@roystgnr do you know what we do for lower D variable output

**Reply by cpgr** [reply]
In the example just above the spikes do go to zero, which is why I suspect an output artefact. I wonder if it is a bit different to the first example.

**Reply by roystgnr** [reply]
I wrote some hacky code to allow users to optionally output fake lower-D elements on which to output `SIDE_HIERARCHIC` variables when using that FE type ... but we're not using fake visualization-only elements here, right?  We're creating actual, part-of-the-mesh-and-assembly lower-D elements.  AFAIK at the libMesh level we don't try to do any sort of projection from higher-D to lower-D elements when we output the latter, we simply output only those variables which are active on that subdomain.

So ... are stress-xx and stress-yy active variables on the lower-D elements?  Do we have something trying to postprocess them that's assuming it's only getting called on a higher-D element?

**Comment by Rahim-Habibi** [comment]
> I wonder if it is just an output artefact that the stress auxvariables aren't projected to the lower dimensional block?

Thanks @cpgr
The issue is not only related to stress projection, also as you can see the displacement behavior sounds inaccurate as well. Actually, the lower dimension at the center moves down more or less "independently" resuting in high displacement zone at the center.

**Comment by Rahim-Habibi** [comment]
> I wrote some hacky code to allow users to optionally output fake lower-D elements on which to output `SIDE_HIERARCHIC` variables when using that FE type ... but we're not using fake visualization-only elements here, right? We're creating actual, part-of-the-mesh-and-assembly lower-D elements. AFAIK at the libMesh level we don't try to do any sort of projection from higher-D to lower-D elements when we output the latter, we simply output only those variables which are active on that subdomain.
> 
> So ... are stress-xx and stress-yy active variables on the lower-D elements? Do we have something trying to postprocess them that's assuming it's only getting called on a higher-D element?

Dear @roystgnr 
yes, it is part-of-the-mesh-and-assembly-lower-D elements. you can see the mesh block at top of the issue. 
regarding the 2nd part of the question, I have checked the input file again, the stresses variable are active on LD. Also, no restriction is set for calling the stress either fo LD or HD. Actually at this step I am trying to initialise the gavity in the model.
